### DISCUSSION TITLE
Physics body not updating with changes in bone rotation of mesh?

### DISCUSSION THREAD

**Discussion by dante-biase** [enable3d/enable3d#196] [Help / Questions] [answered]
I have a rigged mesh loaded via:

```

    var hand
    const loader = new THREE.GLTFLoader().load('./hand.glb', function (gltf) {
        const Object3D = gltf.scene.children[0]

        hand = {
            Object3D: Object3D,
            skinned_mesh: Object3D.children[1],
            skeleton: Object3D.children[1].skeleton,
            position: Object3D.position,
            rotation: Object3D.rotation,
            quaternion: Object3D.quaternion,
        }

        const bones = hand.skeleton.bones;
        hand.forearm = bones[0];
        hand.wrist = bones[1];
        hand.palm = bones[2];
        hand.thumb = bones.slice(3, 7);
        hand.index = bones.slice(7, 11);
        hand.middle = bones.slice(11, 15);
        hand.ring = bones.slice(15, 19);
        hand.pinky = bones.slice(19, 23);

        hand.Object3D.traverse(child => {
            if (child.isMesh) {
                child.castShadow = child.receiveShadow = false
                child.material.metalness = 0
                child.material.roughness = 1
            }
        })
 
        scene.add(hand.Object3D)
        scene.add(new THREE.SkeletonHelper(hand.Object3D))
        physics.add.existing(hand.Object3D, { shape: 'convex', collisionFlags: 2 })
    })
```

it loads as expected into the scene:

![1](https://user-images.githubusercontent.com/45440995/132110784-2fe98fae-b77e-4028-8edb-e7f0b3524857.PNG)

However, when I try to rotate an individual bone on the mesh, the mesh and bones rotate as expected, but the physics body does not remain in sync:

`
hand.palm.rotation.y += Math.PI / 2
`

![2](https://user-images.githubusercontent.com/45440995/132110785-55e326fc-6515-45b4-af56-0daea3ed04d0.PNG)

Is there a way to achieve this?

_____

The documentation says for Kinematic bodies, you need to set `needUpdate` to true when changing rotation or position. I tried setting that after any rotations or transformations via `hand.Object3D.body.needUpdate = true` and this does not seem to fix it either.

I also tried adding `Object3D` to an `new ExtendedObject3D()` as in the headless GLTF example to see if this would make a difference via 

`
Object3D = new ExtendedObject3D()
Object3D.add(gltf.scene.children[0])
`

This does not work and also did not seem necessary anyways because once hand.Object3D is added to physics, it appears it is automatically converted to an ExtendedObject3D.

I'm not sure what else to try besides removing the physics body and then trying to re-add it frame by frame after any transformations? However, this seems horribly inefficient, there must be an already defined approach for doing this that I'm missing.

**Comment by yandeu** [comment, marked as answer]
Does the physics apply correctly if you first rotate the palm?

I have never worked with bones this way.
Can you share your hand model?

---

Maybe all it needs is a simple update somewhere. You should also post this question in the three.js forum. I'm sure they know how to work with bones.

**Reply by dante-biase** [reply]
I cant post this file type here but you can get it from my google drive:

https://drive.google.com/file/d/1O-UbOnM6TFT_0HLUgc7fBM4OywwbpmKl/view?usp=sharing

I definitely will post it on three.js forums and see if anyone else knows what going on as well. I will report back here when if I get some replies.

Is this more related to three.js than enable3d?

**Reply by dante-biase** [reply]
And if you mean does the physics body match the pose if I rotate it before I add it via `physics.add.existing()` no it does not.

However it works rotating the entire model, just not the children/bones

**Reply by yandeu** [reply]
> Is this more related to three.js than enable3d?

I don't know. But we will figure it out 😃
I'm very interested to make it work.

**Reply by yandeu** [reply]
My guess is that you can't deform physics bodies that easily. 

I will check the bullet forum and manual (next week).  
 - https://pybullet.org/Bullet/phpBB3/
- https://github.com/bulletphysics/bullet3/tree/master/docs

**Reply by dante-biase** [reply]
Awesome! Please keep me updated with what you find out. This is the one thing that's preventing me from moving forward on my own project currently. I will post this on those forums today and see if I can find anything in the documentation as well.

**Reply by yandeu** [reply]
I was able to add a soft body. But it was not perfect and run slow.

So I tried adding physics bodies to each bone, which works very well.
See:
- https://youtu.be/lbZFSheTJSE
- https://github.com/enable3d/enable3d/tree/softbody-hand

_Btw, can I keep tha hand and publish an example on the website?_

**Reply by dante-biase** [reply]
hahaha! this looks awesome. The hand slowly creeping up on the box and then flicking I thought was really funny for some reason.

You can absolutely keep the hand and use it as an example for your website. I'm happy this little question ended up contributing something to your project!

One last thing, since the physics bodies are now wrapped around the bones instead of the overall model, I'm wondering if objects could pass through parts of the hand like wrist?

If you think they will is there any way to counter this? Maybe using the mesh as a soft body to wrap around the rigid bodies or ?

Regardless, than you for your help on this! I tried asking in three.js forums and stack overflow but no bites. I'll update this discussion. if I hear anything later on.

**Reply by yandeu** [reply]
You're welcome :)

You would simply adjust the shapes for each bone. Make it fit as much as possible, and no object will pass through.

Btw, check this out: https://youtu.be/e68pu26skLo 😃

**Reply by dante-biase** [reply]
That's awesome. Seems like you've added a huge amount of functionality to your library in such a short time!

Hmm, so you're basically saying manipulate the shape of the bones to fill the mesh with these physics bodies instead of wrapping the entire mesh in physics body because thats only way to achieve the ability to rotate the physics body and mesh body in sync. I wish there was a complete solution, but I will not say no to a workaround for now!

Great library overall by the way. I'd be surprised if this does not pick up more traction than it already has in the near future.

**Reply by yandeu** [reply]
> manipulate the shape of the bones to fill the mesh with these physics bodies instead of wrapping the entire mesh in physics body

No, both work. 
See https://discourse.threejs.org/t/29612/2
But the softBody code is not yet integrated into the framework. It will be in the next release.

_I will make this topic as solved since it will be available soon._

**Reply by dante-biase** [reply]
great! do you mind updating this discussion when it is? This is a critical part of my application

**Reply by AliImrangang** [reply]
we are waiting for this release
